File: src/web/src/models/difference.ts
import { ApiEntry, loadApiEntry } from "./description";

export enum BreakingRank {
    Unknown = -1,
    Compatible = 0,
    Low = 30,
    Medium = 60,
    High = 100,
}

export class DiffEntry {
    id: string = "";
    kind: string = "";
    rank: BreakingRank = BreakingRank.Unknown;
    message: string = "";
    data: any = {};
    old?: ApiEntry;
    new?: ApiEntry;

    from(data: any) {
        this.id = data.id ?? "";
        this.kind = data.kind ?? "";
        this.rank = data.rank ?? BreakingRank.Unknown;
        this.message = data.message ?? "";
        this.data = data.data ?? {};
        if (data.old != undefined) {
            this.old = loadApiEntry(data.old);
        }
        if (data.new != undefined) {
            this.new = loadApiEntry(data.new);
        }
    }
}

File: core/aexpy/models/difference.py
from enum import IntEnum
from typing import Annotated, Any, override

from pydantic import BaseModel, Field

from .description import ApiEntryType


class BreakingRank(IntEnum):
    Unknown = -1
    Compatible = 0
    Low = 30
    Medium = 60
    High = 100


class VerifyState(IntEnum):
    Unknown = 0
    Fail = 50
    Pass = 100


class VerifyData(BaseModel):
    state: VerifyState = VerifyState.Unknown
    message: str = ""
    verifier: str = ""


class DiffEntry(BaseModel):
    id: str = ""
    kind: str = ""
    rank: BreakingRank = BreakingRank.Unknown
    verify: VerifyData = VerifyData()
    message: str = ""
    data: dict[str, Any] = {}
    old: Annotated[ApiEntryType, Field(discriminator="form")] | None = None
    new: Annotated[ApiEntryType, Field(discriminator="form")] | None = None
